libp2p-kad/src/query.rs

```diff
@@ -0,0 +1,335 @@
+            let (message, trigger_idx, other_current_attempts) = match result {
+                Err((err, trigger_idx, other_current_attempts)) => {
+                    (Err(err), trigger_idx, other_current_attempts)
+                }
+                Ok((message, trigger_idx, other_current_attempts)) => {
+                    (Ok(message), trigger_idx, other_current_attempts)
+                }
+            };
+
+            let remote_id = state.current_attempts_addrs.remove(trigger_idx);
+            debug_assert!(state.current_attempts_fut.is_empty());
+            state.current_attempts_fut = other_current_attempts;
+
+            let closer_peers = match message {
+                Ok(closer_peers) => {
+                    // Received a message.
+                    closer_peers
+                }
+                Err(_) => {
+                    state.failed_to_contact.insert(remote_id);
+                    return Ok(future::Loop::Continue(state));
+                }
+            };
+
+            // Update `result` with the node that sent the result.
+            if let Some(insert_pos) = state.result.iter().position(|e| {
+                e.distance_with(&searched_key) >= remote_id.distance_with(&searched_key)
+            }) {
+                if state.result[insert_pos] != remote_id {
+                    state.result.pop();
+                    state.result.insert(insert_pos, remote_id);
+                }
+            } else if state.result.len() < num_results {
+                state.result.push(remote_id);
+            }
+
+            let mut local_nearest_node_updated = false;
+
+            for mut peer in closer_peers {
+                // Update the peerstore with the information sent by
+                // the remote.
+                {
+                    let valid_multiaddrs = peer.multiaddrs.drain(..);
+                    query_interface2.peer_add_addrs(
+                        &peer.node_id,
+                        valid_multiaddrs,
+                        Duration::from_secs(3600),
+                    ); // TODO: which TTL?
+                }
+
+                if peer.node_id.distance_with(&searched_key)
+                    <= state.result[0].distance_with(&searched_key)
+                {
+                    local_nearest_node_updated = true;
+                }
+
+                if state.result.iter().any(|ma| ma == &peer.node_id) {
+                    continue;
+                }
+
+                // Insert the node into `pending_nodes` at the right position, or do not
+                // insert it if it is already in there.
+                if let Some(insert_pos) = state.pending_nodes.iter().position(|e| {
+                    e.distance_with(&searched_key) >= peer.node_id.distance_with(&searched_key)
+                }) {
+                    if state.pending_nodes[insert_pos] != peer.node_id {
+                        state.pending_nodes.insert(insert_pos, peer.node_id.clone());
```

Comment:
As above, if you're reimplementing `BTreeSet` logic why not just use a `BTreeSet`?